File: src/lib/manifest/extractors/fragments.ts
import type { FragmentRecord } from "../types/records";
import type { Extractor, RawTable, RawTableName } from "../types/services";
import type { RawInventoryItem } from "./rawTypes";
import { asRawStatDef } from "./rawTypes";
import { iterItems, projectBase, deriveElement } from "./common";

const FRAGMENT_TYPE_RE = /Fragment$/;
const FRAGMENT_CAT_RE = /\.fragments$/;

/** The six Armor 3.0 stat hashes whose values fragments modify. */
const ARMOR_STAT_HASHES = new Set([
  2996146975, // Mobility
  392767087,  // Resilience
  1943323491, // Recovery
  1735777505, // Discipline
  144602215,  // Intellect
  4244567560, // Strength
]);

type LoadTable = (table: RawTableName) => Promise<RawTable>;

function isFragmentItem(item: RawInventoryItem): boolean {
  const typeName = item.itemTypeDisplayName ?? "";
  const cat = item.plug?.plugCategoryIdentifier ?? "";
  return FRAGMENT_TYPE_RE.test(typeName) || FRAGMENT_CAT_RE.test(cat);
}

function buildStatNameMap(statTable: RawTable): Map<number, string> {
  const map = new Map<number, string>();
  for (const v of Object.values(statTable)) {
    const stat = asRawStatDef(v);
    if (!stat || !stat.displayProperties.name.trim()) continue;
    if (ARMOR_STAT_HASHES.has(stat.hash)) {
      map.set(stat.hash, stat.displayProperties.name);
    }
  }
  return map;
}

function buildStatModifiers(
  item: RawInventoryItem,
  statNameMap: Map<number, string>,
): Record<string, number> {
  const mods: Record<string, number> = {};
  for (const stat of item.investmentStats ?? []) {
    if (stat.isConditionallyActive) continue;
    const name = statNameMap.get(stat.statTypeHash);
    if (name && stat.value !== 0) mods[name] = stat.value;
  }
  return mods;
}

async function extractFragments(loadTable: LoadTable): Promise<FragmentRecord[]> {
  const [itemTable, statTable] = await Promise.all([
    loadTable("DestinyInventoryItemDefinition"),
    loadTable("DestinyStatDefinition"),
  ]);

  const statNameMap = buildStatNameMap(statTable);
  const result: FragmentRecord[] = [];

  for (const item of iterItems(itemTable)) {
    if (!isFragmentItem(item)) continue;

    const typeName = item.itemTypeDisplayName ?? "";
    const cat = item.plug?.plugCategoryIdentifier ?? "";
    const element = deriveElement(typeName, cat);

    result.push({
      ...projectBase(item),
      description: item.displayProperties.description,
      element,
      statModifiers: buildStatModifiers(item, statNameMap),
    });
  }

  return result;
}

export const fragmentsExtractor: Extractor<"fragments"> = {
  store: "fragments",
  extract: (loadTable) => extractFragments(loadTable),
};

File: flutter/packages/manifest/lib/src/extractors/fragments.dart
import '../raw/raw_types.dart';
import '../types/records.dart';
import '../types/services.dart';
import '../types/stores.dart';
import 'common.dart';

final _fragmentTypeRe = RegExp(r'Fragment$');
final _fragmentCatRe = RegExp(r'\.fragments$');

/// Legacy six armor stats used by fragment investment stats in product fixtures.
const _armorStatHashes = <int>{
  2996146975, // Mobility / Weapons (hash reuse across eras)
  392767087, // Resilience / Health
  1943323491, // Recovery / Class
  1735777505, // Discipline / Grenade
  144602215, // Intellect / Super
  4244567560, // Strength
};

class FragmentsExtractor implements EntityExtractor {
  @override
  MvpStoreName get store => MvpStoreName.fragments;

  @override
  Future<List<Object>> extract(LoadRawTable loadTable) async {
    final itemTable = await loadTable('DestinyInventoryItemDefinition');
    final statTable = await loadTable('DestinyStatDefinition');
    final statNameMap = _buildStatNameMap(statTable);
    final result = <FragmentRecord>[];

    for (final item in iterItems(itemTable)) {
      if (!_isFragmentItem(item)) continue;
      final typeName = item.itemTypeDisplayName ?? '';
      final cat = item.plug?['plugCategoryIdentifier'] as String? ?? '';
      final element = deriveElement(typeName, cat);
      final base = projectBase(item);
      result.add(
        FragmentRecord(
          hash: base.hash,
          name: base.name,
          searchName: base.searchName,
          icon: base.icon,
          description: item.displayProperties.description,
          element: element,
          statModifiers: _buildStatModifiers(item, statNameMap),
        ),
      );
    }
    return result;
  }
}

bool _isFragmentItem(RawInventoryItem item) {
  final typeName = item.itemTypeDisplayName ?? '';
  final cat = item.plug?['plugCategoryIdentifier'] as String? ?? '';
  return _fragmentTypeRe.hasMatch(typeName) || _fragmentCatRe.hasMatch(cat);
}

Map<int, String> _buildStatNameMap(RawTable statTable) {
  final map = <int, String>{};
  for (final v in statTable.values) {
    final stat = RawStatDef.tryParse(v);
    if (stat == null || stat.displayProperties.name.trim().isEmpty) continue;
    if (_armorStatHashes.contains(stat.hash)) {
      map[stat.hash] = stat.displayProperties.name;
    }
  }
  return map;
}

Map<String, int> _buildStatModifiers(
  RawInventoryItem item,
  Map<int, String> statNameMap,
) {
  final mods = <String, int>{};
  for (final stat in item.investmentStats) {
    if (stat.isConditionallyActive) continue;
    final name = statNameMap[stat.statTypeHash];
    if (name != null && stat.value != 0) mods[name] = stat.value;
  }
  return mods;
}
